src/node_http2.cc

```diff
@@ -0,0 +1,1326 @@
+              endStream, parent, weight, exclusive);
+
+  nghttp2_priority_spec prispec;
+  nghttp2_priority_spec_init(&prispec, parent, weight, exclusive ? 1 : 0);
+
+  Headers list(isolate, context, headers);
+
+  int32_t ret = session->Nghttp2Session::SubmitRequest(&prispec,
+                                                       *list, list.length(),
+                                                       nullptr, endStream);
+  DEBUG_HTTP2("Http2Session: request submitted, response: %d\n", ret);
+  args.GetReturnValue().Set(ret);
+}
+
+void Http2Session::SubmitResponse(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsNumber());
+  CHECK(args[1]->IsArray());
+
+  Http2Session* session;
+  Nghttp2Stream* stream;
+
+  ASSIGN_OR_RETURN_UNWRAP(&session, args.Holder());
+  Environment* env = session->env();
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+
+  int32_t id = args[0]->Int32Value(context).ToChecked();
+  Local<Array> headers = args[1].As<Array>();
+  bool endStream = args[2]->BooleanValue(context).ToChecked();
+
+  DEBUG_HTTP2("Http2Session: submitting response for stream %d: headers: %d, "
+              "end-stream: %d\n", id, headers->Length(), endStream);
+
+  if (!(stream = session->FindStream(id))) {
+    return args.GetReturnValue().Set(NGHTTP2_ERR_INVALID_STREAM_ID);
+  }
+
+  Headers list(isolate, context, headers);
+
+  args.GetReturnValue().Set(
+      stream->SubmitResponse(*list, list.length(), endStream));
+}
+
+void Http2Session::SubmitFile(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsNumber());  // Stream ID
+  CHECK(args[1]->IsNumber());  // File Descriptor
+  CHECK(args[2]->IsArray());  // Headers
+
+  Http2Session* session;
+  Nghttp2Stream* stream;
+
+  ASSIGN_OR_RETURN_UNWRAP(&session, args.Holder());
+  Environment* env = session->env();
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+
+  int32_t id = args[0]->Int32Value(context).ToChecked();
+  int fd = args[1]->Int32Value(context).ToChecked();
+  Local<Array> headers = args[2].As<Array>();
+
+  DEBUG_HTTP2("Http2Session: submitting file %d for stream %d: headers: %d, "
+              "end-stream: %d\n", fd, id, headers->Length());
+
+  if (!(stream = session->FindStream(id))) {
+    return args.GetReturnValue().Set(NGHTTP2_ERR_INVALID_STREAM_ID);
+  }
+
+  Headers list(isolate, context, headers);
+
+  args.GetReturnValue().Set(stream->SubmitFile(fd, *list, list.length()));
+}
+
+void Http2Session::SendHeaders(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsNumber());
+  CHECK(args[1]->IsArray());
+
+  Http2Session* session;
+  Nghttp2Stream* stream;
+
+  ASSIGN_OR_RETURN_UNWRAP(&session, args.Holder());
+  Environment* env = session->env();
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+
+  int32_t id = args[0]->Int32Value(env->context()).ToChecked();
```

Comment:
I’m just going to mention again that `args[0].As<Int32>()->Value()` is a much less complex operation. I know you’re not a fan of that, but I don’t really see why we wouldn’t want to use it, both from a performance and a code semantics point of view.